### DISCUSSION TITLE
Feedback thread for "Getting Started" docs

### DISCUSSION THREAD

**Discussion by taylorsilva** [concourse/concourse#7353] [Show and tell]
- Did you have any questions after following the tutorial?
- Any sections that were unclear or confusing to follow?
- Other topics or areas of Concourse you think the tutorial should cover?

## 2025 Update

I've updated the Getting Started tutorial and have marked all comments as "outdated" since they all discussed specifics about the now "old" tutorial.

The new tutorial drops the iframes to `localhost:8080` and adds a bunch of images and videos. I hope this better showcases Concourse's web UI.

**Comment by wsierakowski** [comment]
I wonder if the screenshot on this page [tutorial-hello-world](https://concourse-ci.org/tutorial-hello-world.html) is relevant to the topic of this tutorial piece:

![image](https://user-images.githubusercontent.com/2331990/128335635-dd5a31b9-9ae8-47e5-a3b1-6f25623e20af.png)

I would say that instead of 404, we should see a screenshot like below:
![image](https://user-images.githubusercontent.com/2331990/128335780-3f2991fd-f96f-45ba-beae-08a17987141d.png)

**Reply by wsierakowski** [reply]
Obviously, I didn't realise that this wasn't just a screenshot but a frame with a view pointing at the reader's own local instance.

**Reply by taylorsilva** [reply]
Added this line to avoid future confusion: 
<img width="608" alt="image" src="https://user-images.githubusercontent.com/4583837/129761468-bab1ae1c-566e-4b23-a97c-ed59335bffeb.png">

**Comment by roerjo** [comment]
I believe the git commit task needs to be updated to configure git with the email and name. I'm getting an error because of that running it off of the current documentation. I also needed to specify the branch name in the resource configuration for the put step to work. The git resource docs mention that `branch` is required for the put step. https://github.com/concourse/git-resource

**Reply by taylorsilva** [reply]
Thank you! Fixed it https://github.com/concourse/docs/commit/cc0ccd8e9de118c436951dcf5bccd137c6bbbb00

**Reply by roerjo** [reply]
@taylorsilva looks like there are still a couple pieces missing still. Need to call the git config commands still:

```yaml
          params: # Change these to match your personal info
            EMAIL: test@test.com
            GIT_AUTHOR_NAME: Tester
          run:
            path: sh
            args:
              - -cx
              # this is just a bash script
              - |
                cd repo
                date +%Y-%m-%d > todays-date
                git add ./todays-date
                git config --global user.email $EMAIL
                git config --global user.name $GIT_AUTHOR_NAME
                git commit -m "Update todays date"

```

**Reply by taylorsilva** [reply]
I don't think so. I tested with just the environment variables and the task step completed without error. From https://git-scm.com/book/en/v2/Git-Internals-Environment-Variables see the `Committing` section.

**Reply by roerjo** [reply]
Interesting, I kept getting the error until I added the git config commands. I guess either way it will be documented here now. Thanks!

**Reply by taylorsilva** [reply]
Decided to add it just to keep the chance of people running into this error as close to zero as possible :) Thanks for your feedback and going through the tutorial!

**Comment by stranger-danger-zamu** [comment]
This ended up a wall of text going through my first experience using Concourse. There are relevant comments to the tutorial, but also just pain points in general with Concourse. I tried quoting the general pain points and annoyances so the tutorial relevant sections are easier to find.

---

Overall the tutorial was great. It got me up and running fairly quickly and I was able to eventually grok how to put a reasonable pipeline together. I was able to generate artifacts and push them elsewhere by just reading the tutorial. 

It didn't cover how to generate container images so I had to go down a rabbit hole of Github issue spelunking, Github ReadMe's, and searching online for tidbits to figure out how to generate all the artifacts I needed.

Specifically, I was able to generate python `.whl` and `.pyz` files and name them based on the `pyproject.toml` file's version number and push them to an s3 bucket with just the tutorial. When I tried to figure out how to build container images from the artifacts, I spent more than five hours try to find out how to do it idiomatically. 

Eventually, I figured out I had to use `concourse/oci-build-task` to generate the images and did so, but I couldn't figure out how to tag an image when `put`ing it via the `concourse/registry-image` resource. 

> I know I should use the `semver` resource, but that required a backend and I didn't need any of the other functionality it provided and wasn't sure if it would do what I needed it to do without messing with other things. Even then, setting the `semvar` via `put` seemed very complicated for just getting the git repo's matched tag.

> I realized what annoyed me was that some values weren't just added to the variable space even though they were relevant to that particular task (eg. if the git resource is matching against a tag filter, wouldn't it make sense to expose the tag it matched on so it could be used as a variable in that job). 

> I know that resources have a defined interface (check, in, and out) and that in V10 this might be a non-issue, but it's still a pain point now.

At this point I gave up on the idiomatic way and wrangled it together with a custom container that utilized `buildah` to build the images. This was so much easier since I could run commands so I had a lot more control than `oci-build-task`. I even decided to then save the image and `put` the image tagged with `latest` via `registry-image`.

Another reason I went with custom containers was the fact that I couldn't figure out how to pass the self-signed certs to everything. I couldn't find my cert in the task so I could to point to them when running some commands (though resources didn't have cert issues, so it was sorta working? IDK). 

> I know it was supposedly as simple as mounting the certs into `/etc/ssl/certs/<cert>.crt` and it should propagate them to all the workers and containers, but that didn't always work since it didn't propagate them into everything (or at least it didn't copy them into the spawned container for a task, so when I ran buildah it didn't automatically find the cert in `/etc/ssl/certs`). So instead I made more pipelines to construct the custom build containers which embedded the certificate within and used the generated image for other pipelines.

Then I left it running for a bit and eventually it started yelling at me that I hit the max number of workers. But when I look at the resource usage of the host system and I'm idling at most 5% CPU and less than 8% of the 32GiB of memory available. Looking at a process manager, I can see the count of `containerd-shim-runc-v2` rise steadily (with memory increasing, but no CPU usage), so I'm not sure if this is just PEBCAK or an actual issue. 

Overall, Concourse is so much better than other CI's I've tried. After figuring out what I needed to search for it was fairly trivial to get something working, but:

* the fact that a lot of documentation sits on Github ReadMe's is painful
    * searching for answers either points to old Concourse issues (circa v3.x) or points to RFCs or closed issues pointing to an RFC that's still not implemented
* I'm still not sure if docker-compose is a valid installation method or if it's just a kludged together `quickstart`-only method 
* I didn't leave the tutorial understanding why I should use multiple jobs or how to create dependencies between jobs
    * I get that I could use them running test, then building and pushing artifacts, and then smoke testing, and then end to end, but I can also just have them as tasks in a job?
* I couldn't find a way to kill a stuck task
    * This might just be a docker-compose quickstart quirk, a concourse quirk, or a buildah quirk , not sure.
    * If `buildah` is building based on a Dockerfile and it's ambiguous which registry to use (eg. docker.io, quay.io, some internal registry), it'll prompt for user input. This literally causes an impossible to stop task. You can't abort it, cancel it, or anything. I had to blow away the pipeline, land the worker, and restart concourse.

**Comment by chandradeepak** [comment]
→ fly -t tutorial workers
error: unknown target: tutorial

i get this error when i follow the hello-world tutorial

**Reply by taylorsilva** [reply]
Did you `fly login` first? The `login` command creates the target in your `~/.flyrc`

**Comment by dannystaple** [comment]
I am getting an interesting error in running the hello world tutorial:
```
$ fly -t tutorial trigger-job --job hello-world/hello-world-job --watch
started hello-world/hello-world-job #1

initializing
selected worker: ef0e3766321e
find or create container on worker ef0e3766321e: new container: garden spec to oci spec: rootfs must not be empty
errored
```

I've torn down the composed concourse, pruned the volumes and recreated it, and still see this error. Any ideas why?

My (hopefully unsurprising) hello world pipeline file is:
```
jobs:
- name: hello-world-job
  plan:
  - task: hello-world-task
    config:
      platform: linux
      image-resource:
        type: registry-image
        source:
          repository: busybox
      run:
        path: echo
        args: ["Hello concourse!"]
```

I have run the fly commands to login, set-pipeline , unpause and trigger. I have half a TB of disk space remaining - so it's not that. Running on MacOS 11.6.2 with Docker Desktop.

**Reply by dannystaple** [reply]
Doh - found it. My `image-resource` was Kebab-case. It should be snake_case.

**Reply by tylershelton** [reply]
> Doh - found it. My `image-resource` was Kebab-case. It should be snake_case.

One and a half years later, thank you so much for a) teaching me the term "kebab case" and b) pointing out to my blearly, late-night, been-at-this-for-hours eyes that I, too, kebab'd my snake.

**Reply by dannystaple** [reply]
> I, too, kebab'd my snake

That phrase is a keeper...

**Comment by richard-gebbia** [comment]
The base docker-compose file didn't work on my m1 Mac. I had to change the `concourse` service's `image` from `concourse/concourse` to `rdclda/concourse:7.5.0` (which is built for arm architectures).

**Reply by taylorsilva** [reply]
Correct, we don't build an official ARM image of Concourse. I recently got an M1 so I'll be much more motivated to fix this pain sooner rather than later 😁

**Reply by nickmcostello** [reply]
@taylorsilva has this been updated? Is there a native ARM image of Concourse? Could I help get one set up?

**Reply by taylorsilva** [reply]
@nickmcostello sadly it has not been updated. Haven't been spending my spare time on Concourse stuff like I thought I would. Hoping to carve out some time during work hours. Will probably be a few more months until I can make time for that though

**Reply by gsaslis** [reply]
I wish I'd found this thread sooner... 

I've left some relevant comments on https://github.com/concourse/concourse/issues/8719#issuecomment-1676821087 

Even if concourse doesn't ship an `arm64` image, I think a comment in the `docker-compose.yml` that mentions that an `arm64` image is required (it currently directs folks to switch to the `houdini` runtime) would be very helpful! 

(I'd open the PR if I knew where to do that - the link in the quickstart doesn't point to a file in some git repo)

**Reply by taylorsilva** [reply]
With 7.14.0 out, there are official ARM images of Concourse.

**Comment by Kevvix** [comment]
Hi,

I'm trying to do the tutorial but I'm blocked at the part "1.1.2 Hello World Pipeline".
I have been able to do the first part before without problems 

I don't understand the error.

Thanks in advance ^^ 

![image](https://user-images.githubusercontent.com/2465082/181513976-07c0070b-fcef-492a-b7a7-26a435bcbac2.png)

**Reply by taylorsilva** [reply]
Third line from the bottom looks to be the main error message you want to debug. `no such host`. Usually means it was a DNS resolution error. Before that part I can see it's hitting `8.8.8.8` for the DNS lookup, so you already have custom DNS configured. Not sure why you're getting this 🙁 

Is there something up with your Docker networking? Rebooting the docker vm or your computer may resolve this.

**Comment by muzhaqi16** [comment]
Unable to run concourse with apple m1, receiving an error regarding iptables. 
```
error: Exit trace for group:
worker exited with error: Exit trace for group:
garden exited with error: Exit trace for group:
containerd-garden-backend exited with error: setup host network failed: create chain or flush if exists failed: running [/sbin/iptables -t filter -N CONCOURSE-OPERATOR --wait]: exit status 3: iptables v1.6.1: can't initialize iptables table `filter': iptables who? (do you need to insmod?)
Perhaps iptables or your kernel needs to be upgraded.
```

**Comment by paulmiller3000** [comment]
It may be in the docs and I'm just an idiot, but after hours of searching I finally learned that my pipeline cannot pull in a task from another file in memory. For example, I can't get `./scripts/test.sh` and run it as part of my pipeline UNLESS it exists somewhere on the web. For me, this was counterintuitive and represented hours of banging my head against a wall.

Fortunately, I found the statement below at [this site](https://dev.solita.fi/2020/08/17/creating-your-ci-cd-pipeline-with-concourse.html). Making this a bit more clear -- or adding a flashing light for dummies like me -- would be a huge help. And I'll buy you a drink!

> When creating separate task files, it’s important to understand, that only the main pipeline definition will be stored in system memory. Everything else, including separate task definition files, will be fetched from some external resource (e.g. git in our case).

**Comment by Radmanded** [comment]
This isn't a huge problem, but the instructions for installing fly are confusing. In the code block, I copied the commands and the response. Creating a space would help with that. It took me about 10 mins to figure it out, but I'm sure it would stump others. 

$ fly -t tutorial login -c http://localhost:8080 -u test -p test
logging in to team 'main'

<img width="741" alt="CleanShot-Google Chrome202212-09 at 15 10 11@2x" src="https://user-images.githubusercontent.com/9085386/206796836-b29bb814-3602-49a9-a0c7-00dd5f7b0761.png">

Thank you!

**Reply by Radmanded** [reply]
Also, the instructions say to download fly from the web UI. Is it still necessary to use cURL and/or download?

Thanks

**Reply by taylorsilva** [reply]
Feedback incorporated: https://github.com/concourse/docs/pull/537

**Comment by Gr1zzz** [comment]
Hello, i have problem, when i want to connect github. when i started pipeline i have this mistake:
selected worker: 8223477c5283
Cloning into '/tmp/build/get'...
fatal: detected dubious ownership in repository at '/tmp/build/get'
To add an exception for this directory, call:

	git config --global --add safe.directory /tmp/build/get
failed
I alrady update git, and write this command, and write the command for all package save

**Comment by Pashnov** [comment]
I have MacBook Pro, 2021, M1 Pro chip and get next error. Rancher is used instead of docker. Could you please help with it?

`run check: find or create container on worker 76a9e2a62588: starting task: new task: failed to create shim task: OCI runtime create failed: runc create failed: unable to start container process: waiting for init preliminary setup: read init-p: connection reset by peer: unknown`

**Reply by taylorsilva** [reply]
Running docker on ARM machines, like M1, isn't supported right now. You can get it working by setting the runtime to houdini

**Reply by Pashnov** [reply]
how to set runtime as `houdini` ?

**Reply by taylorsilva** [reply]
https://concourse-ci.org/concourse-worker.html#configuring-runtimes

`CONCOURSE_RUNTIME` to `houdini` on the worker. If you're using quickstart then the env var is `CONCOURSE_WORKER_RUNTIME`

**Comment by derynL** [comment]
I am encountering the following 401 error when attempting fly login:
<img width="616" alt="image" src="https://github.com/user-attachments/assets/96736fe8-0800-4174-823f-94f4dc8ed6dc">
Where have I gone wrong? I've followed the preceding steps to the letter.

**Comment by paul-civitas** [comment]
Tutorial is giving a "can not connect to localhost"

<img width="1245" alt="image" src="https://github.com/user-attachments/assets/49df7497-2852-4ab4-bf2e-d65eec437449" />

**Reply by taylorsilva** [reply]
Browser security might be making this too finicky to keep using iframes in our docs. I'm curious what happens if you open [http://localhost:8080/](http://localhost:8080/) in a separate tab.

**Reply by paul-civitas** [reply]
```
This site can’t be reached
localhost refused to connect.
```

**Reply by taylorsilva** [reply]
Ah so that means Concourse didn't even come up. What do you see when you run `docker compose logs` on the Concourse you brought up?

**Reply by paul-civitas** [reply]
Oh I didn't realize the tutorial was supposed to iframe into your localhost lol.

I just skimmed it so I didn't realize it was supposed to be interactive like that and I just assumed it was a bug.

**Reply by taylorsilva** [reply]
LOL fair enough! On the first iframe you run into we do have a little message stating this is an iframe to `localhost:8080`. Might be best to add that everywhere to avoid such confusion.

**Comment by paul-szymanski-a9s** [comment]
I am running the pipeline from https://concourse-ci.org/tutorial-resources.html#put-steps . 
The task step has a run command defined, including this line: `date +%Y-%m-%d > todays-date`.
At some point everything starting from the `>`-sign is dropped.
This is the out put of the job trigger:

```
<snip>
selected worker: 6c0daa98cbf0
running sh -cx cd repo
date +%Y-%m-%d > todays-date
git add ./todays-date
git config --global user.email $EMAIL
git config --global user.name $GIT_AUTHOR_NAME
git commit -m "Update todays date"

+ cd repo
+ date '+%Y-%m-%d'
+ git add ./todays-date
+ git config --global user.email pszymanski@anynines.com
+ git config --global user.name Paul Szymanski
+ git commit -m 'Update todays date'
<snip>
```

What is going on here?

**Reply by paul-szymanski-a9s** [reply]
I seems like there is something in between, that is doing variable substitutions and other things like replacing quotations. What is doing this and why doesn't it play along with `>`?

**Reply by taylorsilva** [reply]
I think everything is working fine, the output is just confusing. The `-cx` part (specifically the `x`) is what tells `sh` to "print every command that you run", but that appears to not include showing redirections of command outputs. So you'll see the command `date` and its args, but not what the output of that command is redirected `>` to.

From the output, I can tell everything is working fine because if it wasn't then the `git add ./todays-date` would fail with `fatal: pathspec 'todays-date' did not match any files`, but it doesn't which means the file was made with the `date '+%Y-%m-%d' > todays-date` command.

I don't think there's anything we can do about this. This is just how `sh` and even `bash` behave when you set `-x`. You can locally, on your laptop, try with this:

```sh
#!/usr/bin/env sh

set -x

echo "hello" > world
cat world
```

When you run that script all you'll see is:
```
+ echo hello
+ cat world
hello
```

You don't see the redirection.

Hope that helps!

**Reply by paul-szymanski-a9s** [reply]
I see, thank you!
I just saw that pipeline actually worked as expected. Previously, I thought this didn't work, because I saw a failed run.
What happened is that the pipeline got rerun after the new resource version was created, which triggers the pipeline again. And the rerun failed, because the date didn't change and therefore there was nothing to commit.
Is there a way to prevent this trigger loop?

**Comment by rudolf-c0** [comment]
Following the [getting started tutorial](https://concourse-ci.org/docs/getting-started/) fails for me because there is no worker node in the docker-compose.yml and the concourse container isn't running in privileged mode so can't start another container. Error message:

```
run check: find or create container on worker 6515ca34311e: starting task: new task: failed to create shim task: OCI runtime create failed: runc create failed: unable to start container process: error during container init: error loading seccomp filter into kernel: error loading seccomp filter: invalid argument: unknown
```

Note: Tried this with concourse 7.12.0 because of local constraints but doubt that running on most recent concourse would change the root cause. Would it be worth adding a worker node to the docker-compose-yml and a section discussing how to set up workers?

**Reply by taylorsilva** [reply]
The concourse container is started with the `quickstart` command, which runs the web and worker in the same container under one process: https://github.com/concourse/docs/blob/baa91cae31eaa095bc10f7aaebe0871e63667caa/docs/docker-compose.yml#L18

The worker must have failed to come up for some reason. You should check the logs for errors with `docker compose logs concourse`.

We do test this compose file in our CI and its currently passing.

What OS/kernel are you running this on? Are you using docker or podman? Any other information you think might be relevant?

**Reply by rudolf-c0** [reply]
Thanks for the reply and I apologize, I missed the "privileged" setting and was running this in an environment that didn't allow privileged execution. 

Since I really want to avoid enabling privileged execution, I am now exploring how to add an external worker. Google and claude are helpful here but it would be nice if there was a docker compose file or tutorial showing this because the way the docker compose file is set up and the way the "Installation" section is structured don't map up very well.

**Reply by taylorsilva** [reply]
There's actually a mostly done set of docs that do go through this! You can view the PR for them here: https://github.com/concourse/docs/pull/603/changes
or click here for a rendered view of the docker compose steps: https://github.com/Aviator-Labs/docs-ci/blob/8b5b85c2269d92f1b59a2029ec7592487304740d/docs/docs/how-to/install-guides/docker-compose.md

Hope that helps!
